AP_Scripting: receive command long and int

IamPete1 opened the pull request:
This is a companion to https://github.com/ArduPilot/ardupilot/pull/13660, in that case we cant selectively listen to a particular command long/int message. This means that AP will go ahead and do what ever and send a ACK. In the case of the user command it will report as unsupported. 

To fix that we have to take the messages and not let AP see them. Obviously it would be bad to just do that for all command longs, so we need to have a special case per command ID, this is that.

I got a little carried away with the buffering, unlike https://github.com/ArduPilot/ardupilot/pull/13660 this can listen to as many commands as it likes from lots of different scripts. 

~~I have also added the ability to send command long and ints to AP from scripting, this is logged as channel -1 (I had to sacrifice the `current` field to get all the log fields to fit, but the spec says its not used)~~

~~As a drive by I have added `#ifdef`  to all the vehicle function only used by scripting, this saves about 500 bytes.~~

~~The only remaining thing is to add the ability to send back ACK's from scripting.~~

Can now send ACKs

IamPete1 commented:
```
Binary           text             data         bss           total
---------------  ---------------  -----------  ------------  ---------------
antennatracker   9356 (+0.7222%)  0 (0.0000%)  4 (+0.0015%)  9360 (+0.6008%)
arducopter       9304 (+0.5557%)  0 (0.0000%)  0 (0.0000%)   9304 (+0.4804%)
arducopter-heli  9376 (+0.5677%)  0 (0.0000%)  0 (0.0000%)   9376 (+0.4899%)
ardurover        9392 (+0.6254%)  0 (0.0000%)  0 (0.0000%)   9392 (+0.5324%)
arduplane        9296 (+0.5548%)  0 (0.0000%)  0 (0.0000%)   9296 (+0.4797%)
blimp            9348 (+0.7651%)  0 (0.0000%)  4 (+0.0015%)  9352 (+0.6301%)
ardusub          9392 (+0.6277%)  0 (0.0000%)  0 (0.0000%)   9392 (+0.5340%)

```
using more flash that I would have guessed...

IamPete1 commented:
```Binary           text             data         bss           total
---------------  ---------------  -----------  ------------  ---------------
antennatracker   6076 (+0.4690%)  0 (0.0000%)  4 (+0.0015%)  6080 (+0.3903%)
arducopter       6024 (+0.3598%)  0 (0.0000%)  0 (0.0000%)   6024 (+0.3111%)
arducopter-heli  6096 (+0.3691%)  0 (0.0000%)  0 (0.0000%)   6096 (+0.3185%)
ardurover        6112 (+0.4070%)  0 (0.0000%)  0 (0.0000%)   6112 (+0.3465%)
arduplane        6016 (+0.3590%)  0 (0.0000%)  0 (0.0000%)   6016 (+0.3104%)
blimp            6068 (+0.4966%)  0 (0.0000%)  4 (+0.0015%)  6072 (+0.4091%)
ardusub          6112 (+0.4085%)  0 (0.0000%)  0 (0.0000%)   6112 (+0.3475%)
```

Fixed the flash issues, got rid of the template class and moved everything over to a cpp. It is still quite allot, I would guess at least half is the new userdata auto generated bindings.

WickedShell commented on libraries/AP_Vehicle/AP_Vehicle.cpp:
```diff
@@ -405,6 +405,43 @@ void AP_Vehicle::publish_osd_info()
 }
 #endif
 
+#ifdef ENABLE_SCRIPTING
```
Don't you need to check if there is also GCS being included in the build? Remember periph has scripting these days but may not have MAVLink.

WickedShell reviewed (changes requested):
I really don't like the injection of commands back from scripting personally. I think it leads to a lot of confusion about what's driving the vehicle to receive a command, and it undermines the point of having proper interfaces for a lot of things. The other thing that's pretty questionable here is that's injecting the command into a random MAVLink stream.

IamPete1 commented:
should add timestamps

IamPete1 commented:
Added timestamps, broken out ifdef changes to a different PR and dropped the sending of command longs/ints. 

This is now revive only. 

We don't actually need the command long/int userdata, we could return the individual components, but that would be a function with 13 returns, and that seems abit much.

WickedShell commented:
> We don't actually need the command long/int userdata, we could return the individual components, but that would be a function with 13 returns, and that seems abit much.

13 returns is in the realm of unreasonably hard to remember what each term is, the userdata gives it all a name for each field, which is vastly more tractable. Technically you could skip the userdata and do a table, but I'm skeptical it's worth the effort to do so, so the userdata's make sense.

IamPete1 commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -4295,6 +4295,14 @@ void GCS_MAVLINK::handle_command_long(const mavlink_message_t &msg)
+#ifdef ENABLE_SCRIPTING
+    AP_Scripting *scripting = AP_Scripting::get_singleton();
+    if (scripting != nullptr && scripting->handle_command_long(packet, chan)) {
+        // Scripting is listening for this command don't process
```
I wonder if we should still log in this case, but we wouldn't know the result of course...

IamPete1 commented:
```
Binary           text             data         bss            total
---------------  ---------------  -----------  -------------  ---------------
antennatracker   5748 (+0.4434%)  0 (0.0000%)  -4 (-0.0015%)  5744 (+0.3685%)
arducopter       5696 (+0.3395%)  0 (0.0000%)  0 (0.0000%)    5696 (+0.2936%)
arducopter-heli  5732 (+0.3465%)  0 (0.0000%)  4 (+0.0015%)   5736 (+0.2993%)
ardurover        5788 (+0.3852%)  0 (0.0000%)  -4 (-0.0015%)  5784 (+0.3277%)
arduplane        5612 (+0.3342%)  0 (0.0000%)  4 (+0.0015%)   5616 (+0.2892%)
blimp            5756 (+0.4707%)  0 (0.0000%)  -4 (-0.0015%)  5752 (+0.3873%)
ardusub          5740 (+0.3833%)  0 (0.0000%)  4 (+0.0015%)   5744 (+0.3264%)
```
Still using lots of flash....

IamPete1 commented:
This now adds a new binding type allowing variable return types. This is the generate binding that results. C++ returns a uint8 bitmask of the returns that are valid. This allows to return both a command long and int from the same function from scriptings point of view. 

```
    mavlink_command_long_t data_5002 = {};
    mavlink_command_int_t data_5003 = {};
    uint32_t data_5004 = {};
    uint8_t data_5005 = {};
    const uint8_t data = static_cast<uint8_t>(ud->receive(
            data_5002,
            data_5003,
            data_5004,
            data_5005));

    if (data & (1U << 0)) {
        new_mavlink_command_long_t(L);
        *check_mavlink_command_long_t(L, -1) = data_5002;
    }
    if (data & (1U << 1)) {
        new_mavlink_command_int_t(L);
        *check_mavlink_command_int_t(L, -1) = data_5003;
    }
    if (data & (1U << 2)) {
        new_uint32_t(L);
        *static_cast<uint32_t *>(luaL_checkudata(L, -1, "uint32_t")) = data_5004;
    }
    if (data & (1U << 3)) {
        lua_pushinteger(L, data_5005);
    }
    return __builtin_popcountll(data);
```

peterbarker commented on libraries/AP_Scripting/AP_Scripting_MAVLink_buffer.cpp:
```diff
@@ -0,0 +1,85 @@
+
+#include "AP_Scripting_MAVLink_buffer.h"
+
+command_buffer::command_buffer(uint8_t buffer_size, int16_t _watch_CMD, uint32_t _watch_msgid) :
```
What's your reasoning on the capitalisation here?

Perhaps `CommandBuffer`?

IamPete1 commented:
Durandal this PR: 

```
Target          Text     Data  BSS     Total  
----------------------------------------------
bin/arducopter  1688972  2564  259796  1951332
```
With no auto userdata bindings saves: 3656
```
bin/arducopter  1685316  2564  259796  1947676
```
With no auto or manual bindings saves another 936
```
bin/arducopter  1684384  2564  259792  1946740
```
master saving another 304
```
bin/arducopter  1684076  2564  259796  1946436
```

So I have the total diff as 4896 (not sure why this is not the same as the 5904 reported in CI) Userdata bindings are using 75% other new bindings are 20% and the rest is 5%.

IamPete1 commented:
Replaced by  https://github.com/ArduPilot/ardupilot/pull/18053
